middle_end/flambda2/to_cmm/to_cmm_env.ml

```diff
@@ -208,150 +282,306 @@ let get_exn_extra_args env k =
           (* Multiple coeffect-only bindings may be accumulated in the same
              stage. *)
-          Coeffect_only (Variable.Map.add var binding bindings) :: stages
+          Coeffect_only (Variable.Set.add var vars) :: stages
         | [] | Effect _ :: _ ->
-          Coeffect_only (Variable.Map.singleton var binding) :: env.stages
+          Coeffect_only (Variable.Set.singleton var) :: env.stages
       in
-      { env with stages }
-
-let bind_variable ?extra env v
-    ~(num_normal_occurrences_of_bound_vars : _ Or_unknown.t)
-    ~effects_and_coeffects_of_defining_expr ~defining_expr =
-  let[@inline] bind_variable0 ~may_inline =
-    bind_variable0 env v ?extra ~effects_and_coeffects_of_defining_expr
-      ~may_inline ~defining_expr
+      { env with stages })
+
+let bind_variable ?extra env var ~defining_expr
+    ~num_normal_occurrences_of_bound_vars
+    ~effects_and_coeffects_of_defining_expr =
+  let inline =
+    To_cmm_effects.classify_let_binding var
+      ~effects_and_coeffects_of_defining_expr
+      ~num_normal_occurrences_of_bound_vars
   in
-  match num_normal_occurrences_of_bound_vars with
-  | Unknown -> bind_variable0 ~may_inline:false
-  | Known num_normal_occurrences_of_bound_vars -> (
-    match
-      To_cmm_effects.classify_let_binding v
-        ~effects_and_coeffects_of_defining_expr
-        ~num_normal_occurrences_of_bound_vars
-    with
-    | Drop_defining_expr -> env
-    | May_inline -> bind_variable0 ~may_inline:true
-    | Regular -> bind_variable0 ~may_inline:false)
+  match inline with
+  | Drop_defining_expr -> env
+  | Regular ->
+    let defining_expr = simple defining_expr in
+    bind_variable_with_decision ?extra env var
+      ~effects_and_coeffects_of_defining_expr ~defining_expr
+      ~inline:Do_not_inline
+  | May_inline_once ->
+    let defining_expr = simple defining_expr in
+    bind_variable_with_decision ?extra env var
+      ~effects_and_coeffects_of_defining_expr ~defining_expr
+      ~inline:May_inline_once
+  | Must_inline_once ->
+    let name = Format.asprintf "%a" Printcmm.expression defining_expr in
+    let defining_expr =
+      complex_no_split name defining_expr effects_and_coeffects_of_defining_expr
+    in
+    bind_variable_with_decision ?extra env var
+      ~effects_and_coeffects_of_defining_expr ~defining_expr
+      ~inline:Must_inline_once
+  | Must_inline_and_duplicate ->
+    let name = Format.asprintf "%a" Printcmm.expression defining_expr in
+    let defining_expr =
+      complex_no_split name defining_expr effects_and_coeffects_of_defining_expr
+    in
+    bind_variable_with_decision ?extra env var
+      ~effects_and_coeffects_of_defining_expr ~defining_expr
+      ~inline:Must_inline_and_duplicate
+
+let bind_variable_to_primitive = bind_variable_with_decision
 
 (* Variable lookup (for potential inlining) *)
 
-let will_inline env binding = binding.cmm_expr, env, binding.effs
+let split_complex_binding (binding : complex binding) =
```

Comment:
This function might be clearer with either a custom datatype for its return type of a comment specifying what the `None` and `Some _` cases mean.